File: Others/BOJ_10825_2.java
//조건 변형
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Scanner;

public class Main {
	static class Student { // Student 구조체를 정의합니다
		String name;
		int kor;
		int eng;

		public Student(String name,int kor,int eng) {
			this.name = name;
			this.kor = kor;
			this.eng = eng;
		}
	}
    public static ArrayList<String> solution(int N, String [] name, int [] kor, int [] eng){
    	ArrayList <String> answer = new ArrayList<>();
      ArrayList <Student> list = new ArrayList<>(); // Student들을 담을 ArrayList를 생성합니다
    	for(int i = 0; i < N; i++) {
			list.add(new Student(name[i],kor[i],eng[i])); 
		}
    	Collections.sort(list,new Comparator<Student>() { // 정렬합니다
			@Override
			public int compare(Student o1, Student o2) {
				if(o1.kor==o2.kor) {
					if(o1.eng==o2.eng) {
							return o1.name.compareTo(o2.name);
					}return o2.eng-o1.eng;
				}else
					return o1.kor-o2.kor;
			}
		});
    	for(Student s : list) answer.add(s.name);
    	return answer;
    }
	public static void main(String[] args){
	  Scanner sc = new Scanner(System.in);
	  StringBuilder sb = new StringBuilder();
		int N = sc.nextInt();
		String [] name = new String[100010];
		int [] kor = new int [100010];
		int [] eng = new int [100010];

		for(int i = 0; i < N; i++) {
			name[i] = sc.next();
			kor[i] = sc.nextInt();
			eng[i] = sc.nextInt();
		}
		ArrayList<String> answer = solution(N,name,kor,eng);
		for(String s : answer) sb.append(s).append("\n");
		System.out.print(sb);
		sc.close();
	}
}

File: Others/BOJ_10825_2.py
#조건 변형
import sys
input = sys.stdin.readline  

def solution(n, name, kor, eng):
    arr = [(name[i], kor[i], eng[i]) for i in range(n)]
    # sort()가 2차원 혹은 그 이상의 리스트를 정렬할 때, 기본적으로 리스트의 첫번째 요소부터 오름차순으로 비교를 합니다.
    # 따라서 정렬의 우선순위를 변경하기 위해서는 key값에 람다 함수를 작성합니다.
    # 람다 함수는 비교 순서대로 작성하되, 내림차순이 필요한 경우 음수를 붙여줍니다.
    arr.sort(key=lambda x:(x[1], -x[2], x[0]))


    # 정렬된 리스트에서 이름 가져오기
    answer = [arr[i][0] for i in range(n)]
    return answer

# 입출력을 포함한 메인 코드
if __name__ == "__main__":
    # 입력
    n = int(input())
    kor = [0] * n
    eng = [0] * n
    name = [''] * n
    for i in range(n):
        name[i], scores = input().split(' ', 1)
        kor[i], eng[i] = map(int, scores.split())
    
    # 연산
    answer = solution(n, name, kor, eng)
    
    # 출력
    print(*answer, sep='\n')